lib/logstash/filters/kubernetes_metadata.rb

```diff
@@ -0,0 +1,288 @@
+
+    event.set(@target, metadata)
+    return filter_matched(event)
+  end
+
+  def set_log_formats(metadata)
+    begin
+      #return if metadata['annotations'].empty?
+
+      format = {
+        'stderr' => @default_log_format,
+        'stdout' => @default_log_format
+      }
+      a = metadata['annotations']
+      n = metadata['container_name']
+
+      # check for log-format-<stream>-<name>, log-format-<name>, log-format-<stream>, log-format
+      # in annotations
+      %w{ stderr stdout }.each do |t|
+        [ "log-format-#{t}-#{n}", "log-format-#{n}", "log-format-#{t}", "log-format" ].each do |k|
+          if v = a[k]
+            format[t] = v
+            break
+          end
+        end
+      end
+
+      metadata['log_format_stderr'] = format['stderr']
+      metadata['log_format_stdout'] = format['stdout']
+      @logger.debug("kubernetes metadata => #{metadata}")
+
+    rescue => e
+      @logger.warn("Error setting log format: #{e}")
+    end
+  end
+
+  # based on https://github.com/vaijab/logstash-filter-kubernetes/blob/master/lib/logstash/filters/kubernetes.rb
+  def get_file_info(path)
+    parts = path.split(File::SEPARATOR).last.gsub(/.log$/, '').split('_')
+    if parts.length != 3 || parts[2].start_with?('POD-')
+      return nil
+    end
+    kubernetes = {}
+    kubernetes['replication_controller'] = parts[0].gsub(/-[0-9a-z]*$/, '')
+    kubernetes['pod'] = parts[0]
+    kubernetes['namespace'] = parts[1]
+    kubernetes['container_name'] = parts[2].gsub(/-[0-9a-z]*$/, '')
+    kubernetes['container_id'] = parts[2].split('-').last
+    return kubernetes
+  end
+
+  def sanatize_keys(data)
+    return {} unless data
+
+    parsed_data = {}
+    data.each do |k,v|
+      new_key = k.gsub(/\.|,/, '_')
+        .gsub(/\//, '-')
+      parsed_data[new_key] = v
+    end
+
+    return parsed_data
+  end
+
+  def get_kubernetes(namespace, pod)
+
+    begin
+      @logger.debug("Attempting to query the Kubernetes API")
+
+      url = [ @api, 'api/v1/namespaces', namespace, 'pods', pod ].join("/")
+
+      rest_opts = {
+        verify_ssl: false
+      }
+
+      if @auth
+        if @auth['basic']
+          @logger.debug("Found basic auth for Kubernetes API")
+
+          basic_user = @auth['basic']['user']
+          basic_pass = @auth['basic']['pass']
+
+          rest_opts.merge!( user: basic_user, password: basic_pass )
+        end
+
+        if @auth['bearer']
+          @logger.debug("Found Bearer  auth for Kubernetes API")
+
+          bearer_key = @auth['bearer']['key']
+
+          rest_opts.merge!( Authorization: "Bearer #{bearer_key}" )
+        end
+      end
+
+      @logger.debug("rest_opts: #{rest_opts}")
+
+      begin
+        response = RestClient::Resource.new(url, rest_opts).get
+      rescue RestClient::ResourceNotFound
+        @logger.warn("Kubernetes returned an error while querying the API")
+        @logger.warn("url: #{url}, rest_opts: #{rest_opts}")
```

Comment:
I recommend not logging `rest_opts` because it can contain secrets (user, password, tokens, etc), and we should not log secrets.